keps/sig-storage/20190709-csi-snapshot.md

```diff
@@ -89,28 +92,150 @@ The rest of the document includes required information missing from the original
+         DeletionPolicy DeletionPolicy
+         Driver string
+         VolumeSnapshotClassName *string
+}
+```
+
+```
+type VolumeSnapshotContentSource struct {
+	// +optional
+	VolumeHandle *string
+	// +optional
+	SnapshotHandle *string
+}
+```
+
+```
+type VolumeSnapshotContentStatus struct {
+  CreationTime *int64
+  ReadyToUse *bool
+  RestoreSize *int64
+  Error *VolumeSnapshotError
+  SnapshotHandle *string
+}
+```
+
+The beta Kubernetes VolumeSnapshotClass API object is as follows:
+
+```
+type VolumeSnapshotClass struct {
+        metav1.TypeMeta
+        metav1.ObjectMeta
+
+        Driver string
+        Parameters map[string]string
+        DeletionPolicy DeletionPolicy
+}
+```
+
+### Controller Split
+
+When Volume Snapshot is promoted to Beta in Kubernetes 1.17, the CSI external-snapshotter sidecar controller is split into two controllers: a snapshot-controller and a CSI external-snapshotter sidecar.
+
+The snapshot controller will be watching the Kubernetes API server for `VolumeSnapshot` and `VolumeSnapshotContent` CRD objects. The CSI `external-snapshotter` sidecar only watches the Kubernetes API server for `VolumeSnapshotContent` CRD objects.
+
+The creation of a new `VolumeSnapshot` object referencing a `SnapshotClass` CRD object corresponding to this driver causes the snapshot controller to trigger the creation of a Kubernetes `VolumeSnapshotContent` object to represent the to-be-created new snapshot.
+
+The creation of a new `VolumeSnapshotContent` object causes the sidecar container to trigger a `CreateSnapshot` operation against the specified CSI endpoint to provision a new snapshot. When a new snapshot is successfully provisioned, the sidecar container updates the status field of the `VolumeSnapshotContent` object to represent the new snapshot.
+
+The snapshot controller will be updating the status field of the `VolumeSnapshot` object accordingly based on the status field of the `VolumeSnapshotContent` object to indicate the new snapshot is ready to be used.
+
+The deletion event of a `VolumeSnapshot` object bound to a `VolumeSnapshotContent` corresponding to this driver with a `delete` deletion policy causes the snapshot controller to start deleting the `VolumeSnapshotContent` object and add an annotation to the object to indicate it is being deleted. Note that both the `VolumeSnapshot` object and the `VolumeSnapshotContent` object will not be deleted immediately due to the finalizers. When the sidecar container detects this update on the `VolumeSnapshotContent` object, it triggers a `DeleteSnapshot` operation against the specified CSI endpoint to delete the snapshot. Once the snapshot is successfully deleted, the sidecar container removes the finalizer on the `VolumeSnapshotContent` object which leads to the deletion of the object from Kubernetes. The snapshot controller then removes the finalizer on the `VolumeSnapshot` object and as a result the object will be deleted from Kubernetes.
```

Comment:
I am not sure how much details we need to expose in this type of documents, do we need to cover all major cases? for example, what if the deletion policy is retain?
@msau42 @xing-yang

Reply:
I added a sentence to briefly describe what will happen if the deletion policy is retain.

Reply:
I think it should be fine as long as we have detailed documentation in the external-snapshot repo. In this KEP probably its OK to keep it on high level.
I'd suggest a nit to add a small piece to elaborate what to expect if a user deletes a bound "VolumeSnapshotContent" object directly, it will have a deletion timestamp set however will persist in API server until its corresponding "VolumeSnapshot" object gets deleted.

Reply:
Done